File: yggdrasil/examples/formatted_io4/src/formatted_io4_modelA.cpp
#include <iostream>
// Include methods for input/output channels
#include "YggInterface.hpp"

int main(int argc, char *argv[]) {
  // Initialize input/output channels
  YggAsciiArrayInput in_channel("inputA");
  YggAsciiArrayOutput out_channel("outputA", "%6s\t%ld\t%f\n");

  // Declare resulting variables and create buffer for received message
  size_t nrows;
  int flag = 1;
  char *name = NULL;
  long *count = NULL;
  double *size = NULL;

  // Loop until there is no longer input or the queues are closed
  while (flag >= 0) {
  
    // Receive input from input channel
    // If there is an error, the flag will be negative
    // Otherwise, it is the size of the received message
    flag = in_channel.recvRealloc(4, &nrows, &name, &count, &size);
    if (flag < 0) {
      std::cout << "Model A: No more input." << std::endl;
      break;
    }

    // Print received message
    printf("Model A: (%lu rows)\n", nrows);
    size_t i;
    for (i = 0; i < nrows; i++)
      printf("   %.6s, %ld, %f\n", &name[6*i], count[i], size[i]);

    // Send output to output channel
    // If there is an error, the flag will be negative
    flag = out_channel.send(4, nrows, name, count, size);
    if (flag < 0) {
      std::cout << "Model A: Error sending output." << std::endl;
      break;
    }

  }
  
  // Free dynamically allocated columns
  if (name) free(name);
  if (count) free(count);
  if (size) free(size);
  
  return 0;
}

File: yggdrasil/examples/formatted_io4/src/formatted_io4_modelA.f90
program main
  ! Include methods for input/output channels
  use fygg

  ! Declare resulting variables and create buffer for received message
  logical :: flag = .true.
  type(yggcomm) :: in_channel, out_channel
  integer(kind=c_size_t), target :: nrows
  type(character_1d), target :: name
  type(c_long_1d), target :: count
  type(real8_1d), target :: siz
  integer(kind=c_size_t) :: i

  ! initialize input/output channels
  in_channel = ygg_ascii_array_input("inputA")
  out_channel = ygg_ascii_array_output("outputA", "%6s\t%ld\t%f\n")

  ! Loop until there is no longer input or the queues are closed
  do while (flag)

     ! Receive input from input channel
     ! If there is an error, the flag will be negative
     ! Otherwise, it is the number of variables filled
     flag = ygg_recv_var_realloc(in_channel, [ &
          yggarg(nrows), yggarg(name), yggarg(count), yggarg(siz)])
     if (.not.flag) then
        print *, "Model A: No more input."
        exit
     end if

     ! Print received message
     print *, "Model A: (", nrows, " rows)"
     do i = 1, nrows
        print *, "   ", name%x(i)%x, count%x(i), siz%x(i)
     end do

     ! Send output to output channel
     ! If there is an error, the flag will be negative
     flag = ygg_send_var(out_channel, [ &
          yggarg(nrows), yggarg(name), yggarg(count), yggarg(siz)])
     if (.not.flag) then
        print *, "Model A: Error sending output."
        exit
     end if

  end do

end program main
